polygeist/include/mlir/Dialect/Polygeist/Utils/TransformUtils.h

```diff
@@ -53,19 +53,50 @@ class IfThenElseBodyInterface {
 // Loop Versioning Utilities
 //===----------------------------------------------------------------------===//
 
+/// Represents a loop versioning condition.
+class LoopVersionCondition {
+public:
+  using SCFCondition = Value;
+
+  struct AffineCondition {
+    IntegerSet ifCondSet;
+    SmallVectorImpl<Value> &setOperands;
+  };
+
+  /// Create a loop versioning condition for SCF loops.
+  LoopVersionCondition(SCFCondition scfCond)
+      : versionCondition({scfCond, std::nullopt}) {}
+
+  LoopVersionCondition(AffineCondition affineCond)
+      : versionCondition({std::nullopt, affineCond}) {}
+
+  SCFCondition getSCFCondition() const {
+    assert(versionCondition.scfCond.has_value() &&
+           "expecting valid SCF condition");
+    return *versionCondition.scfCond;
+  }
+  AffineCondition getAffineCondition() const {
+    assert(versionCondition.affineCond.has_value() &&
+           "expecting valid affine condition");
+    return *versionCondition.affineCond;
+  }
+
+private:
+  struct VersionCondition {
```

Comment:
We could use `std::variant` instead.

Reply:
I was initially thinking that some consumer could create a versioning condition holding both a valid  `SCFCondition` and an `AffineCondition` but I think that is unlikely and probably overkill. I'll make the change to use a std::variant.